packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4157,3 +4164,244 @@ class _CopySelectionAction extends ContextAction<CopySelectionTextIntent> {
   @override
   bool get isActionEnabled => state._value.selection.isValid && !state._value.selection.isCollapsed;
 }
+
+/// A void function that takes a [TextEditingValue].
+@visibleForTesting
+typedef TextEditingValueCallback = void Function(TextEditingValue value);
+
+/// Provides undo/redo capabilities for text editing.
+///
+/// Listens to [controller] as a [ValueNotifier] and saves relevant values for
+/// undoing/redoing. The cadence at which values are saved is a best
+/// approximation of the native behaviors of a hardware keyboard on Flutter's
+/// desktop platforms, as there are subtle differences between each of these
+/// platforms.
+///
+/// Listens to keyboard undo/redo shortcuts and calls [onChanged] when a
+/// shortcut is triggered that would affect the state of the [controller].
+@visibleForTesting
+class TextEditingHistory extends StatefulWidget {
+  /// Creates an instance of [TextEditingHistory].
+  const TextEditingHistory({
+    Key? key,
+    required this.child,
+    required this.controller,
+    required this.onChanged,
+  }) : super(key: key);
+
+  /// The child widget of [TextEditingHistory].
+  final Widget child;
+
+  /// The [ValueNotifier] to save the state of over time.
+  final TextEditingController controller;
+
+  /// Called when an undo or redo is received, even if the state would be the
+  /// same.
+  final TextEditingValueCallback onChanged;
+
+  @override
+  State<TextEditingHistory> createState() => _TextEditingHistoryState();
+}
+
+// State for [TextEditingHistory].
```

Comment:
This comment is redundant

Reply:
Good call, I think I was thinking this was public and needed a comment.